transformations/transformations/global_var_offload.py

```diff
@@ -0,0 +1,275 @@
+        for pragma in FindNodes(Pragma).visit(routine.body):
+            if pragma.keyword == 'loki':
+                if 'update_device' in pragma.content:
+                    if update_device:
+                        pragma_map[pragma] = update_device
+                    else:
+                        pragma_map[pragma] = None
+
+                if 'update_host' in pragma.content:
+                    if update_host:
+                        pragma_map[pragma] = update_host
+                    else:
+                        pragma_map[pragma] = None
+
+        routine.body = Transformer(pragma_map).visit(routine.body)
+
+        # build new imports to add offloaded global vars to driver symbol table
+        new_import_map = {}
+        for s in self._var_set:
+            if s in routine.symbol_map:
+                continue
+
+            if new_import_map.get(self._modules[s], None):
+                new_import_map[self._modules[s]] += as_tuple(s)
+            else:
+                new_import_map.update({self._modules[s]: as_tuple(s)})
+
+        new_imports = ()
+        for k, v in new_import_map.items():
+            new_imports += as_tuple(Import(k, symbols=tuple(Variable(name=s, scope=routine) for s in v)))
+
+        # add new imports to driver subroutine sepcification
+        import_pos = 0
+        if (old_imports := FindNodes(Import).visit(routine.spec)):
+            import_pos = routine.spec.body.index(old_imports[-1]) + 1
+        if new_imports:
+            routine.spec.insert(import_pos, Comment(text=
+               '![Loki::GlobalVarOffload].....Adding global variables to driver symbol table for offload instructions'))
+        import_pos += 1
+        for index, i in enumerate(new_imports):
+            routine.spec.insert(import_pos + index, i)
+
+    def process_kernel(self, routine, **kwargs):
+        """
+        Collect the set of module variables to be offloaded.
+        """
+
+        # filter out procedure imports
+        calls = [c.name for c in FindNodes(CallStatement).visit(routine.body)]
+        calls += FindInlineCalls().visit(routine.body)
+
+        imports = [i for i in FindNodes(Import).visit(routine.spec) if i.symbols not in calls]
+
+        # build map of modules corresponding to imports
+        import_mod = {s.name.lower(): i.module for i in imports for s in i.symbols}
+
+        # separate out derived and basic types
+        basic_types = [s.name.lower() for i in imports for s in i.symbols if s in kwargs['targets']
+                       if isinstance(s.type.dtype, BasicType) and s.name.lower() in self._var_set]
+        deriv_types = [s for i in imports for s in i.symbols if s in kwargs['targets']
+                       if isinstance(s.type.dtype, DerivedType) and s.name.lower() in self._var_set]
+
+        with dataflow_analysis_attached(routine):
+
+            # collect symbols to add to acc update pragmas in driver layer
+            for item in basic_types:
```

Comment:
Just a matter of taste: I would not use `item` here, as it is a somewhat reserved name in transformations to refer to the `Item` object

Reply:
Agreed, names now changed.